crates/wysiwyg/src/composer_model/menu_state.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::dom::nodes::ContainerNodeKind;
+use crate::dom::Range;
+use crate::menu_state::MenuStateUpdate;
+use crate::{
+    ComposerModel, DomHandle, DomNode, ListType, MenuState, ToolbarButton,
+    UnicodeString,
+};
+
+impl<S> ComposerModel<S>
+where
+    S: UnicodeString,
+{
+    pub(crate) fn compute_menu_state(&mut self) -> MenuState {
+        let (s, e) = self.safe_selection();
+        let range = self.state.dom.find_range(s, e);
+        match range {
+            Range::SameNode(range) => {
+                return self.compute_active_buttons(range.node_handle);
+            }
+            Range::MultipleNodes(range) => {
+                return self.compute_active_buttons(
+                    // TODO: lookup all locations and output union instead
```

Comment:
I'm not actually sure what the right behaviour is here - we should compare with other editors.

Reply:
I've self reviewed up here ^ I meant intersection, from looking at some others editors, same rule applies to list than the one for formatting. 
Button is active if and only if the entire selection is inside a list of expected type
For example, no active button with this:
<img width="583" alt="image" src="https://user-images.githubusercontent.com/80891108/189076149-1517a33a-dd30-4b8d-a164-c2ac74ac4f6b.png">

Reply:
I think the button state should exactly match what will happen then I click it.
So if the button is not highlighted, clicking it will add (or replace) a list, and if it is highlighted, clicking it will remove a list.

Reply:
Will have to check what happens when this is implemented for us.
In my screenshot here if I hit one of the list buttons, it will transform the "unmatching" list to the same type and merge which looks like a correct behaviour to me.